File: 2022/07/30/remove_anagrams.py
"""
https://leetcode.com/problems/find-resultant-array-after-removing-anagrams/
Author: Andrew Jarombek
Date: 7/30/2022
"""

from typing import List


def remove_anagrams(words: List[str]) -> List[str]:
    if len(words) == 1:
        return words

    codes = ["".join(sorted(list(word))) for word in words]

    index = 1

    while index < len(words):
        if codes[index] == codes[index - 1]:
            del codes[index]
            del words[index]
        else:
            index += 1

    while len(words) > 1 and codes[-1] == codes[-2]:
        del codes[-1]
        del words[-1]

    return words


if __name__ == '__main__':
    assert remove_anagrams(["abba", "baba", "bbaa", "cd", "cd"]) == ["abba", "cd"]

File: 2022/10/11/removeAnagrams.ts
/**
 * https://leetcode.com/problems/find-resultant-array-after-removing-anagrams/
 * @author Andrew Jarombek
 * @since 10/11/2022
 */

function removeAnagrams(words: string[]): string[] {
    const result = [];
    let index = 0;
    let prev = '';

    while (index < words.length) {
        let chars = words[index].split('');
        chars.sort((a, b) => a.charCodeAt(0) - b.charCodeAt(0));
        const key = chars.join('');

        if (prev != key) {
            prev = key;
            index++;
        } else {
            words.splice(index, 1);
            index = 0;
            prev = '';
        }
    }

    return words;
}

function removeAnagramsV1(words: string[]): string[] {
    const result = [];
    const set = new Set<string>();

    for (let word of words) {
        let chars = word.split('');
        chars.sort((a, b) => a.charCodeAt(0) - b.charCodeAt(0));
        const key = chars.join('');

        if (!set.has(key)) {
            result.push(word);
            set.add(key);
        }
    }

    return result;
}